cookbooks/cdo-varnish/templates/default/accept-language.vcl.erb

```diff
@@ -0,0 +1,188 @@
+    /* We have to copy, otherwise root_lang will be the same every time */ \
+    strncpy(pl[curr_lang].lang, x, LANG_MAXLEN); \
+    pl[curr_lang].q = y;       \
+    curr_lang++;               \
+}
+
+struct lang_list {
+    vcl_string lang[LANG_MAXLEN];
+    float q;
+};
+
+/* In-place lowercase of a string */
+static void strtolower(char *s) {
+    register char *c;
+    for (c=s; *c; c++) {
+        if (isupper(*c)) {
+            *c = tolower(*c);
+        }
+    }
+    return;
+}
+
+/* Checks if a given language is in the static list of the ones we support */
+int is_supported(vcl_string *lang) {
+    vcl_string *supported_languages = SUPPORTED_LANGUAGES;
+    vcl_string match_str[LANG_MAXLEN + 3] = "";  /* :, :, \0 = 3 */
+    int is_supported = 0;
+
+    /* We want to match 'zh-cn' and 'zh-CN' too */
+    strtolower(lang);
+
+    /* Search ":<lang>:" in supported languages string */
+    strncpy(match_str, ":", 1);
+    strncat(match_str, lang, LANG_MAXLEN);
+    strncat(match_str, ":\0", 2);
+
+    if (strstr(supported_languages, match_str))
+        is_supported = 1;
+
+    return is_supported;
+}
+
+/* Used by qsort() below */
+int sort_by_q(const void *x, const void *y) {
+    struct lang_list *a = (struct lang_list *)x;
+    struct lang_list *b = (struct lang_list *)y;
+    if (a->q > b->q) return -1;
+    if (a->q < b->q) return 1;
+    return 0;
+}
+
+/* Reads Accept-Language, parses it, and finds the first match
+   among the supported languages. In case of no match,
+   returns the default language.
+*/
+void select_language(const vcl_string *incoming_header, char *lang) {
+
+    struct lang_list pl[LANG_LIST_SIZE];
+    vcl_string *lang_tok = NULL;
+    vcl_string root_lang[3];
+    vcl_string *header;
+    vcl_string header_copy[HDR_MAXLEN];
+    vcl_string *pos = NULL;
+    vcl_string *q_spec = NULL;
+    unsigned int curr_lang = 0, i = 0;
+    float q;
+
+    /* Empty or default string, return default language immediately */
+    if (
+        !incoming_header
+        || (0 == strcmp(incoming_header, "en-US"))
+        || (0 == strcmp(incoming_header, "en-GB"))
+        || (0 == strcmp(incoming_header, DEFAULT_LANGUAGE))
+        || (0 == strcmp(incoming_header, ""))
+    )
+        RETURN_DEFAULT_LANG;
+
+    /* Tokenize Accept-Language */
+    header = strncpy(header_copy, incoming_header, sizeof(header_copy));
+
+    while ((lang_tok = strtok_r(header, " ,", &pos))) {
+
+        q = 1.0;
+
+        if ((q_spec = strstr(lang_tok, ";q="))) {
+            /* Truncate language name before ';' */
+            *q_spec = '\0';
+            /* Get q value */
+            sscanf(q_spec + 3, "%f", &q);
+        }
+
+        /* Wildcard language '*' should be last in list */
+        if ((*lang_tok) == '*') q = 0.0;
+
+        /* Push in the prioritized list */
+        PUSH_LANG(lang_tok, q);
+
+        /* For cases like 'en-GB', we also want the root language in the final list */
+        if ('-' == lang_tok[2]) {
```

Comment:
I believe this could crash varnish by dereferencing off the end of the string strlen(lang_tok) < 2-.  They don't check the length of the string and unconditionally reference lang_tok[2[.